### ISSUE TITLE
<random>: Floating-point random number generation suboptimal due to generate_canonical()

### ISSUE THREAD

**Issue by statementreply** [microsoft/STL#1000]
**Describe the bug**

The function template `generate_canonical()` makes a call to `std::log2()` to determine the number of calls to the random number generator given (in order to acquire enough entropy):
https://github.com/microsoft/STL/blob/5be7d49c243979231dff35919d3a3d813d75d714/stl/inc/random#L268-L273

Although the number of calls required (`_Kx`) is a compile-time constant, the call to `std::log2()` cannot be optimized away even if /fp:fast is specified so that every time we generate a floating-point random number through `std::generate_canonical()`, which is the case for all the `std::*_distribution`'s instantiated for a floating-point type, we must pay for a cost of a function call to `std::log2()`.

Also, there is another cost for (inlined) `std::ceil()` followed by a floating-point to int cast.

The incurred cost of totally unnecessary function calls and a cast can be significant for such an application that makes heavy use of, say, a randomized algorithm.

**STL version**

```
https://github.com/microsoft/STL/commit/5be7d49c243979231dff35919d3a3d813d75d714
```

**Additional context**

Also tracked by DevCom-717452 and Microsoft-internal VSO-976509 / [AB#976509](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/976509).

**Comment by MattStephanson** [comment]
I wasn't aware of this issue at the time, but it seems that #1964 is a duplicate of this and is fixed by #2498.

**Comment by CaseyCarter** [comment]
Closing as duplicate of #1964. Thanks, @MattStephanson.
